### DISCUSSION TITLE
Overloading a TypeGuard function (pd.isnull and related variants)

### DISCUSSION THREAD

**Discussion by gandhis1** [microsoft/pyright#3995] [General]
From the original issue here: https://github.com/pandas-dev/pandas-stubs/pull/339. It was recommended that I bring this to the discussion forum here to solicit some feedback, and if this in fact a gap in the spec, potentially escalating this to the typing-sig list. 

In summary, the linked PR is an attempt of mine to create type annotations for the function `pd.isnull` and all related variants in `pandas`. This function has varying behavior based on what is passed to it. I'm not going to exhaustively list all overloads, but as an example, if you pass ` pd.DataFrame` to it, it will return a `pd.DataFrame`. If you pass a regular Python object, or a scalar such as a `Optional[int]`, it will return a `bool`. It does not only consider `None` as null, it also considers several other types, including `pd.NA`, and `pd.NaT`. My goal was to enhance the existing type annotations such that it can be used as a type guard, while still keeping the existing behavior: when an object guaranteed to be non-null or guaranteed to be null is passed to it, it returns a `Literal[True]` or `Literal[False]`.

The rest of the context I am basically quoting verbatim below. So my questions essentially are:

1) Should the spec for `TypeGuard` support overloaded functions? This use case I think makes a compelling case for it.
2) Should overlapping overloads be allowed, and it uses the narrowest one? Or should a `TypeGuard` be able to decompose `Union` parameters such that passing a partial overlap of the `Union` will maintain the narrowest type? Perhaps https://mail.python.org/archives/list/typing-sig@python.org/thread/EMUD2D424OI53DCWQ4H5L6SJD2IXBHUL/ is relevant

**Except from the linked issue:**

***

The limiting factor I see right now is that yes, `pd.isna` has a variable return type, and `TypeGuard` always needs to return `bool`. But this seems to me a glaring gap in the capability of the type checker - we should be able to narrow types.

If I have this code, which is essentially the same as the test I added:

```python
x = random.choice("value", None)  # Optional[str]
if pd.notnull(x):
     ...  # x is guaranteed to be a str
```

We know for a fact that x is a `str` in the body of the `if` statement. I am not aware of any exceptions to that. Anyway, rather than giving up here due to type checking limitations, I'm wondering if this is something worth escalating to the typing mailing list. Is the restriction that `TypeGuard` always returns a `bool` actually necessary? Or can we support `@overload` with `TypeGuard`?

The second, somewhat separate issue, is how the narrowing works. I have a test case illustrating this as well. But to copy the example, this first block of code already works as intended and passes MyPy at least:

```python
x = random.choice("value", None, pd.NA)  # Union[str, None, pd.NA]
if pd.isnull(x):
    ...  # x is Union[None, pd.NA]
```

The above code works because I annotated all of the null-equivalent types in a single `Union`. However, if you do that, then this following example does NOT work. It should work. But it doesn't:

```python
x = random.choice("value", None)  # Union[str, None]
if pd.isnull(x):
    ...  # x should be None, but because we had a single TypeGuard overload that covered pd.NA as well, this is actually Union[None, pd.NA]
```

Now the question I was asking myself, so is the solution to just have separate overloads for each? But you can't, because that overlaps with the original broader overload, and *pyright* complains about overlapping overloads.

**Comment by erictraut** [comment]
The `TypeGuard` feature is independent of `overload`. Both features are useful, and they are intended to compose.

I'm a bit confused by your question and the samples because they involve incomplete code snippets. It would be helpful if you could put together a self-contained minimal code sample that provides the basis for your question.

Perhaps what you're looking for is the `StrictTypeGuard` feature that I've proposed (and implemented in prototype form). It's a variant of PEP 647's TypeGuard that is less flexible than the existing form but allows for narrowing in the negative ("else") case and narrowing of unions. For details, refer to [this typing-sig thread](https://mail.python.org/archives/list/typing-sig@python.org/thread/EMUD2D424OI53DCWQ4H5L6SJD2IXBHUL/) and this [python/typing thread](https://github.com/python/typing/discussions/1013#discussioncomment-1966238).

**Reply by gandhis1** [reply]
I guess to start, how can they compose if `TypeGuard` are by spec required to only ever return a `bool`? The function in question `pandas.isnull` does not do that.

**Reply by erictraut** [reply]
`TypeGuard` applies to individual overload signatures, like in the following.

```python
from typing import TypeGuard, overload

@overload
def is_str(x: str) -> TypeGuard[str]: ...

@overload
def is_str(x: object) -> list[bool]: ...

def func(a: str, b: list[str]):
    if is_str(a):
        reveal_type(a) # str
    
    if is_str(b):
        reveal_type(b) # list[str]
```

In this example, `is_str` returns a `bool` in some cases, and it returns a `list[bool]` in others. In the case where it returns a `bool`, `TypeGuard` can be used (with the associated type narrowing behaviors). In the case where it returns `list[bool]`, `TypeGuard` cannot be used, so type narrowing isn't supported in that case.

**Reply by gandhis1** [reply]
From PEP 647:

> TypeGuard is a special form that accepts a single type argument. It is used to annotate the return type of a user-defined type guard function. Return statements within a type guard function should return bool values, and **type checkers should verify that all return paths return a bool**.

How is that overload example consistent with that?

**Reply by erictraut** [reply]
That sentence in the PEP is referring to the validation of a TypeGuard function's _implementation_, not the signature. I admit that when I wrote that part of the PEP, I didn't have overload implementations in mind. So I can see why you would be confused about how it might apply to an overload implementation. But it works fine if you use a `TypeGuard` within a union for the overload implementation's return type.

```python
def is_str(x: object) -> TypeGuard[str] | list[bool]:
    if isinstance(x, str):
        return True
    if isinstance(x, list):
        return [is_str(y) for y in x]
    return False
```

Pyright and mypy are both fine with this.

**Reply by gandhis1** [reply]
Ok, so that's good news, I was mis-interpreting that sentence. So the second part of my question is annotating this:

```import random
from typing_extensions import TypeGuard
from typing import Union
from pandas._libs.missing import NAType


def isnull(x: Union[str, None, NAType]) -> TypeGuard[Union[None, NAType]]:
    if x is None or x is NAType:
        return True
    else:
        return False


x = random.choice(["value", None])
reveal_type(x)  #  Type of "x" is "str | None"
if isnull(x):
    reveal_type(x)  # Type of "x" is "NAType | None"
```

The goal is for the second `reveal_type` to be `None` only. How can I do this? If I add 3 overloads, one for `str | None`, one for `str | NAType`, and keep the original `str | None | NAType`, then it becomes an overlapping overload.

**Reply by erictraut** [reply]
> How can I do this?

That can't be done today with `TypeGuard`. You would need `StrictTypeGuard` to make that work.

**Comment by gandhis1** [comment]
Appreciate all the responses. The `StrictTypeGuard` feature is exactly what I'm looking for. I look forward to that making it into the spec someday. My PR here must unfortunately use a less accurate annotation in its absence: https://github.com/pandas-dev/pandas-stubs/pull/339
